charger/daheimladen.go

```diff
@@ -1,214 +1,351 @@
+	dhlRegChargedEnergy   = 72  // Uint16 RO 0.1kWh
+	dhlRegChargingTime    = 78  // Uint32 RO 1s
+	dhlRegSafeCurrent     = 87  // Uint16 WR 0.1A
+	dhlRegCommTimeout     = 89  // Uint16 WR 1s
+	dhlRegCurrentLimit    = 91  // Uint16 WR 0.1A
+	dhlRegChargeControl   = 93  // Uint16 WR ENUM
+	dhlRegChargeCmd       = 95  // Uint16 WR ENUM
+	dhlRegVoltages        = 108 // 3xUint32 RO 0.1V
+
+	dhlCmdStartSession = 1
+	dhlCmdStopSession  = 2
+)
+
 func init() {
 	registry.Add("daheimladen", NewDaheimLadenFromConfig)
 }
 
 // NewDaheimLadenFromConfig creates a DaheimLaden charger from generic config
 func NewDaheimLadenFromConfig(other map[string]interface{}) (api.Charger, error) {
-	cc := struct {
-		Token     string
-		StationID string
-		Cache     time.Duration
-	}{
-		Cache: time.Second,
+	cc := modbus.TcpSettings{
+		ID: 255,
 	}
 
 	if err := util.DecodeOther(other, &cc); err != nil {
 		return nil, err
 	}
 
-	return NewDaheimLaden(cc.Token, cc.StationID, cc.Cache)
+	return NewDaheimLaden(cc.URI, cc.ID)
 }
 
 // NewDaheimLaden creates DaheimLaden charger
-func NewDaheimLaden(token, stationID string, cache time.Duration) (*DaheimLaden, error) {
-	c := &DaheimLaden{
-		Helper:      request.NewHelper(util.NewLogger("daheim")),
-		stationID:   stationID,
-		connectorID: 1,
-		idTag:       daheimladen.EVCC_IDTAG,
-		token:       token,
-		cache:       cache,
+func NewDaheimLaden(uri string, id uint8) (api.Charger, error) {
+	conn, err := modbus.NewConnection(uri, "", "", 0, modbus.Tcp, id)
+	if err != nil {
+		return nil, err
 	}
 
-	c.Client.Transport = &oauth2.Transport{
-		Source: oauth2.StaticTokenSource(&oauth2.Token{
-			AccessToken: token,
-			TokenType:   "Bearer",
-		}),
-		Base: c.Client.Transport,
-	}
+	log := util.NewLogger("dhl")
+	conn.Logger(log.TRACE)
 
-	c.statusCache = provider.ResettableCached(func() (daheimladen.GetLatestStatus, error) {
-		var res daheimladen.GetLatestStatus
-		err := c.GetJSON(fmt.Sprintf("%s/cs/%s/status", daheimladen.BASE_URL, c.stationID), &res)
-		return res, err
-	}, c.cache)
+	wb := &DaheimLaden{
+		log:  log,
+		conn: conn,
+		curr: 60, // assume min current
+	}
 
-	c.meterCache = provider.ResettableCached(func() (daheimladen.GetLatestMeterValueResponse, error) {
-		var res daheimladen.GetLatestMeterValueResponse
-		err := c.GetJSON(fmt.Sprintf("%s/cs/%s/metervalue", daheimladen.BASE_URL, c.stationID), &res)
-		return res, err
-	}, c.cache)
+	// get failsafe timeout from charger
+	b, err := wb.conn.ReadHoldingRegisters(dhlRegCommTimeout, 1)
+	if err != nil {
+		return nil, fmt.Errorf("failsafe timeout: %w", err)
+	}
+	go wb.heartbeat(time.Duration(binary.BigEndian.Uint16(b)/2) * time.Second)
 
-	return c, nil
+	return wb, err
 }
 
-// reset cache
-func (c *DaheimLaden) reset() {
-	c.statusCache.Reset()
-	c.meterCache.Reset()
+func (wb *DaheimLaden) heartbeat(timeout time.Duration) {
+	for range time.NewTicker(timeout).C {
+		if _, err := wb.conn.ReadInputRegisters(dhlRegSafeCurrent, 1); err != nil {
```

Comment:
@premultiply Changing the register address to `dhlRegConnectorState` seems to get rid of errors coming from every heartbeat

Reply:
My fault. Looks like simple Copy&Paste error. Should be `ReadHoldingRegisters()` ...

Reply:
Should be fixed now. Please check again.